pkg/assets/builder.go

```diff
@@ -127,32 +148,117 @@ func (a *AssetBuilder) RemapImage(image string) (string, error) {
 	}
 
 	a.ContainerAssets = append(a.ContainerAssets, asset)
-
 	return image, nil
 }
 
 // RemapFile sets a new url location for the file, if a AssetsLocation is defined.
-func (a AssetBuilder) RemapFile(file string) (string, error) {
+func (a *AssetBuilder) RemapFileAndSHA(file string, sha string) (string, *hashing.Hash, error) {
+	if file == "" {
+		return "", nil, fmt.Errorf("unable to remap an empty string")
+	}
+
+	fileAsset := &FileAsset{
+		File: file,
+		SHA:  sha,
+	}
+
+	if a.AssetsLocation != nil && a.AssetsLocation.FileRepository != nil {
+		fileAsset.CanonicalLocation = file
+		fileAsset.CononicalSHALocation = sha
+
+		file, err := a.normalizeURL(file)
+		if err != nil {
+			return "", nil, fmt.Errorf("unable to parse file url %q: %v", file, err)
+		}
+
+		fileAsset.File = file
+
+		sha, err = a.normalizeURL(sha)
+		if err != nil {
+			return "", nil, fmt.Errorf("unable to parse sha url %q: %v", file, err)
+		}
+
+		fileAsset.SHA = sha
+		glog.V(4).Infof("adding remapped file: %+v", fileAsset)
+	}
+
+	h, err := a.FindHash(fileAsset)
+	if err != nil {
+		return "", nil, err
+	}
+
+	fileAsset.SHAValue = h.Hex()
+	a.FileAssets = append(a.FileAssets, fileAsset)
+
+	glog.V(8).Infof("adding file: %+v", fileAsset)
+
+	return fileAsset.File, h, nil
+}
+
+// RemapFileAndSHAValue is used exclusively to remap the cni tarball, as the tarball
+// does not have a sha file in object storage.
+func (a *AssetBuilder) RemapFileAndSHAValue(file string, shaValue string) (string, error) {
 	if file == "" {
 		return "", fmt.Errorf("unable to remap an empty string")
 	}
 
 	fileAsset := &FileAsset{
-		File:              file,
-		CanonicalLocation: file,
+		File:     file,
+		SHAValue: shaValue,
 	}
 
 	if a.AssetsLocation != nil && a.AssetsLocation.FileRepository != nil {
-		fileURL, err := url.Parse(file)
+		fileAsset.CanonicalLocation = file
+
+		file, err := a.normalizeURL(file)
 		if err != nil {
 			return "", fmt.Errorf("unable to parse file url %q: %v", file, err)
 		}
 
-		fileRepo := strings.TrimSuffix(*a.AssetsLocation.FileRepository, "/")
-		fileAsset.File = fileRepo + fileURL.Path
+		fileAsset.File = file
+		glog.V(4).Infof("adding remapped file: %q", fileAsset.File)
 	}
 
 	a.FileAssets = append(a.FileAssets, fileAsset)
 
 	return fileAsset.File, nil
 }
+
+// FindHash returns the hash value from remove sha file via https.
+func (a *AssetBuilder) FindHash(file *FileAsset) (*hashing.Hash, error) {
```

Comment:
Not clear to me why we're not checking SHA / CanonicalSHALocation first here?

Does this function need to be public?

Reply:
It is a two phase process. First during assest phase the original location is checked.  Then during the other phases the canonical location is checked.